### PR TITLE
Added dot and cos merge modes

### PR DESCRIPTION
**Pull request by farizrahman4u**
Two useful merge modes: dot product merge and cosine merge.

### CODE REVIEW

**Comment by dbonadiman** [comment]
The way i had implemented the shape is not elegant at all :D try something like that instead             

```python
 elif self.mode =='dot':
       return tuple(input_shapes[0][0], 1)
```

File: keras/layers/core.py [hunk 1]
```diff
@@ -298,6 +300,10 @@ def output_shape(self):
             return tuple(output_shape)
         elif self.mode == 'join':
             return None 
+        elif self.mode in ['dot','cos']:
```
Comment (dbonadiman) [inline keras/layers/core.py]: The way i had implemented the shape is not elegant at all :D try something like that instead             

```python
 elif self.mode =='dot':
       return tuple(input_shapes[0][0], 1)
```

File: keras/layers/core.py [hunk 2]
```diff
@@ -263,12 +263,14 @@ def get_config(self):
...
         ''' Merge the output of a list of layers or containers into a single tensor.
-            mode: {'sum', 'mul', 'concat', 'ave', 'join'}
+            mode: {'sum', 'mul', 'concat', 'ave', 'join', 'dot', 'cos'}
         '''
         if len(layers) < 2:
             raise Exception("Please specify two or more input layers (or containers) to merge")
-        if mode not in {'sum', 'mul', 'concat', 'ave', 'join'}:
+        if mode not in {'sum', 'mul', 'concat', 'ave', 'join', 'dot', 'cos'}:
             raise Exception("Invalid merge mode: " + str(mode))
+        if mode in ['dot', 'cos'] and len(layers) > 2:
```
Comment (dbonadiman) [inline keras/layers/core.py]: The two 'vectors should have the same shape too'

**Comment by farizrahman4u** [comment]
@dbonadiman  I included shape checking.. But is it necessary? sum. average, elementwise multiplication all requires vectors of same shape.. but shape checking is not done for any of them. Should not shape checking implemented in general for all these modes? Like:

if mode in ['sum','ave','mul','cos','dot']:
   shape=layers[0].output_shape
   for l in layers:
       if l.output_shape != shape:
           raise Exception(mode + " merge requires layers of same output shape")

If yes, then it should be a separate PR

**Comment by farizrahman4u** [comment]
Done. Shape checking has been proposed in a separate PR #881 . 
Removing shape checking for dot and cos in this one.

**Comment by iskandr** [comment]
Does the "dot" merge mode cover the case where one side is reshaped to a m x n matrix and the other side is a length n vector? I've found this occasionally useful (though it would scale to larger matrices if we had some way to represent them in a factorized form).

**Comment by farizrahman4u** [comment]
@iskandr  No it doesnt, as of now. Infact, I am waiting for #881 to be merged to add strict shape checking. What you are proposing could be easily accomplished, once this PR is merged. Simply repeat the n vector using RepeatVector layer m times to get a mxn model, this can then be merged with your other mxn model using dot or cos mode.

**Comment by fchollet** [comment]
This is neat. But I think we would need to be able to specify the axis of the merge. For instance, with timeseries (3D tensors), sometimes you will want to dot the features (axis=2), and sometimes you will want to dot the time data points (axis=1).

So I see a need for an `axis` keyword argument. For a cleaner API, we could remove the `concat_axis` argument and use the same `axis` argument for all merge modes that use an axis.

Also, I believe we can implement these new merge modes without `scan`.

**Comment by farizrahman4u** [comment]
@fchollet  The default value of ```concat_axis``` is ```-1```, whereas the default value for axis that makes sense for ```dot``` merge is ```2```. What do we do about this? What about:
```python
if mode == 'dot':
    if axis == -1:
         axis = 2
```
But it doesnt "feel" right.

**Comment by farizrahman4u** [comment]
Sorry, I screwed up. Automatically merge-able PR : #891

**Comment by fchollet** [comment]
Right. The [Numpy API](http://docs.scipy.org/doc/numpy/reference/generated/numpy.tensordot.html) allows the `axes` argument to be an int, a tuple of 2 lists / arrays of same length. In this context it's clearer to use a different keyword.

**Pull request closed without merging**